### ISSUE TITLE
ChaCha20: change the current API (and recommendations)

### ISSUE THREAD

**Issue by woodruffw** [pyca/cryptography#9017]
This is a tie-together for Cryptography's side of the discussions in https://github.com/google/wycheproof/issues/90 and https://bugs.chromium.org/p/boringssl/issues/detail?id=614.

## Status quo

The current `ChaCha20` API in Cryptography is defined as follows:

```python
algorithms.ChaCha20(key, nonce)
```

...where `nonce` is a `bytes` of exactly `len(nonce) == 16`. This means that it contains both the "nonce" and "counter" fields used by ChaCha20, and leaves it to the underlying implementation to determine the appropriate nonce/counter split.

As part of that, the current documentation also encourages uses to fully randomize the `nonce` input, with `os.urandom(16)`:

https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#cryptography.hazmat.primitives.ciphers.algorithms.ChaCha20

Finally, the API implies that the underlying `ChaCha20` implementation implementation is RFC 7539, when the two current underlying implementations (OpenSSL and LibreSSL) are not (due to a 64/64 nonce/ctr split rather than 96/32).

Put together, these three points mean that the current `ChaCha20` API's internal implementations are more likely to diverge from RFC 7539 than they otherwise would be: if a user randomly initializes `nonce`, then the randomly initialized interior counter may be much closer to RFC 7539's block limit than a user might otherwise expect. 

I have two proposals for resolving this: one that forces all implementations to behave as if they're RFC 7539 compliant, and another that explicitly separates the API into 64/64 and 96/32 implementations.

## Proposal 1

My first proposal is a semi-breaking change. Behaviorally:

1. The API itself will change, with `nonce` becoming a 12-byte nonce rather than a 16-byte nonce;
2. Internally, the `ChaCha20` API will initialize the 32-bit counter (to 0) and combine it with the nonce.
3. _Optionally_, the API could grow a new `counter=` argument, which would allow the user to explicitly initialize the counter to a specific value in `[0, 2^32)`. 

In effect, this would make the non-RFC-compliant implementations of ChaCha20 behave as if they're compliant within the well-defined parameters of the RFC (i.e., up to 256 GB of enciphering), and would ensure that the majority of users never exceed those parameters (since most users will not need to manually initialize the counter).

Notably, this would be a breaking API change in terms of public API parameters, but a non-breaking change in terms of *documented* behavior (since it would be a bugfix to enforce compliance with RFC 7539, as documented). 

### Proposal 2

My second proposal is non-breaking, but requires a documentation correction and a new API for the RFC 7539-compliant version of ChaCha20. Behaviorally:

1. The existing ChaCha20 API's documentation will be updated to emphasize that it isn't intended to be RFC 7539 compliant. We'd also add test cases to ensure compliance with the 64/64 variant, e.g. at what would be the 32-bit counter overflow point. 
2. We'd add a new `ChaCha20RFC` API (with a better name than that), which would then support the RFC 7539 variant. This API would then use the pre-existing `ChaCha20` API with a fixed 32-bit nonce, forcing all implementations to behave as if they're compliant (at least within the bounds defined by the RFC -- exceeding those bounds will still exhibit divergences).

This would be non-breaking. However, if a _little_ bit of breakage is acceptable, I think combining this with a small change to `ChaCha20` to make `nonce` only 8 bytes (and allow an optional 64-bit counter) would be the best of both worlds: doing so would eliminate any confusion over initial high counter values, and would make the RFC-compliant version's interior use of the non-RFC version slightly more explicit.

See also #8956.

cc @facutuesca

**Comment by ronf** [comment]
RFC 7539 isn’t the only spec that this code needs to be able to support. In particular, Chacha20Poly1305 shows up in SSH (see the documentation at https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.chacha20poly1305 from OpenSSH). That relies on a 64/64 split on the unique nonce vs. counter value, and also relies on being able to initialize the counter explicitly to specific byte string values (like 0 and little-endian 1). So, I’m not sure it is a good idea to enforce a specific split or other constraints on the allowed counter values. The current API leaves this up to the callers, and I think we’d want to maintain that as at least one of the API options.

If you wanted to add a higher-level API which built on this and provided specific splits and automatic counter initialization to simplify things for a caller looking for something like RFC 7539 support, I’d be ok with that, but I would suggest doing it as a very thin layer on the existing API which takes an already constructed 128 bit value.

**Comment by woodruffw** [comment]
> RFC 7539 isn’t the only spec that this code needs to be able to support. In particular, Chacha20Poly1305 shows up in SSH (see the documentation at https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.chacha20poly1305 from OpenSSH). That relies on a 64/64 split on the unique nonce vs. counter value, and also relies on being able to initialize the counter explicitly to specific byte string values (like 0 and little-endian 1).

Yep, this was a consideration behind the second proposal above: I didn't phrase it especially well, but the idea there would be to retain the 64/64 behavior with the existing `ChaCha20` API, and then have an entirely new API (either a class, or a constructor that does the split correctly) that does the RFC-specified variant.

The more general problem here is that Cryptography's various backends don't all agree on which variant they implement :slightly_smiling_face: -- OpenSSL and LibreSSL appear to have intentionally (?) implemented the 64/64 variant, while BoringSSL provides the 96/32 RFC variant.

**Comment by davidben** [comment]
Being able to specify starting nonce values definitely makes sense. That's a good way to reserve a couple points in the counter space for the MAC key or whatever. The IETF AEAD does that too. It's specifically setting it to random values and relying on wraparound that's incoherent.

**Comment by bitloi** [comment]
@alex Can I pick this issue up?

**Comment by alex** [comment]
This issue still requires significant design decisions, it's not a good task for a new contributor.

**Comment by reaperhulk** [comment]
https://github.com/pyca/cryptography/pull/14977

cc @ronf. I believe this won't break you since OpenSSH's structure of packet sequences should never overflow a 32-bit counter, but I want to confirm that 😄

**Comment by ronf** [comment]
> cc @ronf. I believe this won't break you since OpenSSH's structure of packet sequences should never overflow a 32-bit counter, but I want to confirm that 😄

Yes - the packet sequence number that is used to construct the nonce is always a 32-bit value which starts at zero (and resets back to 0 when beginning to encrypt the channel when both sides support "strict key" checks). It is possible for this sequence number to wrap, but typically re-keying would be done long before that occurs (defaulting to every 1 GiB of data or 1 hour of time).

**Comment by ronf** [comment]
One thing that worried me is the comment "ChaCha20 now extracts the 32-bit little-endian block counter from the first 4 bytes of the 128-bit nonce."

Looking at SSH's encoding of the 128-bit nonce, the first 8 bytes are a little-endian "counter" value of 0 or 1 and this is followed by an 8 byte nonce which holds a big-endian sequence number which will never exceed 32 bits. However, this sequence number would not be the first 4 bytes of the nonce. It would be the last 4 bytes of the 16 byte (128-bit) counter plus nonce. Given that the split between the counter and nonce portion can also vary, I'm not sure it is a good idea to assume anything about the layout of this 128-bit value.

**Comment by reaperhulk** [comment]
I'm not sure I follow. From the draft RFC for OpenSSH with ChaCha20Poly1305:

> The next 2 words x[12],x[13] are a 64-bit counter which is incremented for each output data block, with x[12] being the low-order 32 bits of the counter and x[13] the high-order bits. The final 2 words x[14],x[15] are initialized from an 8-byte "nonce" which must be a unique identifier for each separate message encrypted with the same key.

So in this construction you'd still pass your 128-bit value that you've constructed, but the low order 32-bits would be treated as the counter. Since OpenSSH will never encrypt 256GiB without changing the packet sequence or entirely rekeying then this should be safe. Unless I'm wrong about what OpenSSH does here!

**Comment by ronf** [comment]
From https://datatracker.ietf.org/doc/html/draft-josefsson-ssh-chacha20-poly1305-openssh-00:

```
   The AEAD is constructed as follows: for each packet, generate a
   Poly1305 key by taking the first 256 bits of ChaCha20 stream output
   generated using K_2, an IV consisting of the packet sequence number
   encoded as an uint64 under the SSH wire encoding rules and a ChaCha20
   block counter of zero.  The K_2 ChaCha20 block counter is then set to
   the little-endian encoding of 1 (i.e. {1, 0, 0, 0, 0, 0, 0, 0}) and
   this instance is used for encryption of the packet payload.
```

Note the "uint64 under the SSH wire encoding rules". SSH wire encoding is a big-endian encoding, so the 32-bit SSH packet sequence number is encoded as a big-endian 64-bit value (4 bytes of zeroes followed by the 32-bit sequence number in bigendian format). So, if you are expecting the counter to be the first 4 bytes and to be little-endian, that would be the reverse of what the API is getting right now.

https://datatracker.ietf.org/doc/draft-ietf-sshm-chacha20-poly1305/01/ doesn't have as much to say about the byte order of the nonce -- it just requires that it must be unique. However,  the nonce as encoding in OpenSSH and AsyncSSH today is big-endian, and it follows the (little-endian) 64-bit counter value that alternates between 0 & 1 for encrypting packet headers vs. data. The resulting 128-bit value is what's passed in each encrypt/decrypt call.

Is it possible that there's a endian-swap of the 16 byte counter+nonce internal to the PyCA internal implementation that's not captured here?

**Comment by alex** [comment]
I don't think those docs are inconsistent with this. Looking at the OpenSSH source code: https://github.com/openssh/openssh-portable/blob/master/cipher-chachapoly-libcrypto.c#L90-L130 it looks like for the Poly1305 key it uses the sequence number for the second 64-bits, with a 0 counter, and then sets counter to 1 (clearly little endian).

The asyncssh code appears to do the same: https://github.com/ronf/asyncssh/blob/develop/asyncssh/crypto/chacha.py#L37-L46

**Comment by ronf** [comment]
The relevant code in AsyncSSH around encoding the nonce is https://github.com/ronf/asyncssh/blob/develop/asyncssh/encryption.py#L221-L247, where it uses UInt64() to convert packet sequence numbers to big-endian byte strings. See also https://github.com/ronf/asyncssh/blob/develop/asyncssh/packet.py#L62-L65.

Basically, SSH wire encoding for integers is big-endian. The counter values here are little-endian as you said, but that's an exception, and isn't using standard SSH encoding for those. The end result is a mixture of big and little endian data in the final 128-bit counter+nonce.

**Comment by ronf** [comment]
Actually, looking more closely at the change, I see that the limit you're placing is actually on the counter portion of the counter+nonce, but that will only ever be 0 or 1 (little-endian) in the case of SSH, so we may be good here. The endianness of the nonce looks like it won't matter to this code, and even the fact that it is a 64-bit counter value and 64-bit nonce (instead of 32 & 96 bits) shouldn't end up mattering here.

**Comment by reaperhulk** [comment]
Right, that's the core of what we're trying to make sure we're all aligned on. If we all agree that for the OpenSSH scenario the 32-bit little endian counter can never overflow then we can start enforcing this "breaking" change while having a great deal of confidence that the only breaking component would be for consumers who encrypted data with a counter value that would have overflowed the 32-bit counter.

@alex let's talk a bit tomorrow on this before we merge, I have one other thought on this 😄
